### PR TITLE
Setup running phoenix tests with OCIS

### PR DESCRIPTION
**Pull request by dpakach**
#### Setup running phoenix ui acceptance tests using the OCIS bakend.
Currently we run all of the acceptance tests of phoenix using the OC10 owncloud bakcend. But Phoenix will also need to work with the new OCIS backend. This PR sets up all the required configuration for running the phoenix tests with the ocis backend. As of now Most of the features from OC10 are missing in OCIS. Therefore most tests will fail while running on the OCIS backend. But with the setup of running tests with OCIS, we can run the tests as the development process goes on which will help developers to track and avoid bugs that may occur later.

## Related Issue

- Related https://github.com/owncloud/phoenix/issues/1995

## How Has This Been Tested?

- locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

File: tests/acceptance/helpers/ldapHelper.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+const ldap = require('ldap')
+const { join } = require('./path')
+const fs = require('fs-extra')
+
+const userHelper = require('./userSettings')
+
+exports.createClient = function (url) {
+  return new Promise((resolve, reject) => {
+    const ldapClient = ldap.createClient({
+      url: url || client.globals.ldap_url // 'ldap://127.0.0.1'
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]:
```suggestion
      url: url || client.globals.ldap_url
```

File: tests/acceptance/helpers/ldapHelper.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    })
+
+    ldapClient.bind('cn=admin,dc=owncloud,dc=com', 'admin', err => {
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: should be configurable see e.g. https://github.com/owncloud/core/pull/36791/files#diff-d8f5321aed99cdfd86ee188d13edb329R85

File: tests/acceptance/helpers/ldapHelper.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  return exports.createEntry(ldapClient, `uid=${user},ou=TestUsers,dc=owncloud,dc=com`, {
+    cn: user.charAt(0).toUpperCase() + user.slice(1),
+    sn: user,
+    objectclass: ['posixAccount', 'inetOrgPerson'],
+    homeDirectory: `/home/openldap/${user}`,
+    userPassword: password,
+    displayName: displayName,
+    mail: email,
+    gidNumber: 5000,
+    uidNumber: 1
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: needs to be unique for every user

File: tests/acceptance/helpers/ldapHelper.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  }).then(err => {
+    if (!err) {
+      userHelper.addUserToCreatedUsersList(user, password, displayName, email)
+      const skelDir = process.env.OCIS_SKELETON_DIR
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: other env. variables are processed in nightwatch.js why not this one?

File: tests/acceptance/stepDefinitions/provisioningContext.js [hunk 5]
```diff
@@ -143,6 +146,9 @@ function addToGroup (userId, groupId) {
 }
 
 Given('user {string} has been created with default attributes', function (userId) {
+  if (client.globals.ocis) {
```
Comment (individual-it) [inline tests/acceptance/stepDefinitions/provisioningContext.js]: what about putting this logic into `createDefaultUser` ?

File: tests/acceptance/helpers/ldapHelper.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+const userHelper = require('./userSettings')
+
+exports.createClient = function (url) {
+  return new Promise((resolve, reject) => {
+    const ldapClient = ldap.createClient({
+      url: url || client.globals.ldap_url
+    })
+
+    ldapClient.bind(client.globals.ldap_base_dn, 'admin', err => {
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: suggesting to put the password also into a configurable env-var

File: tests/acceptance/helpers/ldapHelper.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      if (err) {
+        reject(err)
+      }
+    })
+
+    ldapClient.add('ou=TestUsers,dc=owncloud,dc=com', {
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: OU for groups and users is repeated multiple times, please put into a const

File: tests/acceptance/helpers/ldapHelper.js [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  })
+}
+
+exports.getNewUID = function (ldapClient) {
+  const userCn = 'ou=TestUsers,dc=owncloud,dc=com'
+  let uid = 1
+  return new Promise((resolve, reject) => {
+    ldapClient.search(
+      userCn,
+      { scope: 'sub', attributes: ['uidNumber'] },
```
Comment (individual-it) [inline tests/acceptance/helpers/ldapHelper.js]: does this get always the last (highest) id on the server?

Comment (dpakach) [reply tests/acceptance/helpers/ldapHelper.js]: Yes, It checks all the users and gets the new id from them (The new id will be max id + 1 )

**Comment by dpakach** [comment]
Blocked by https://github.com/owncloud/ocis-reva/issues/58. Without custom redirect uris we cannot run the tests in CI. Tests run locally but few of them pass as of now.

**Comment by individual-it** [comment]
maybe unblocked by https://github.com/owncloud/ocis-phoenix/pull/41 ?

**Comment by butonic** [comment]
@dpakach how is this blocked? the CI tests are green.

I have this error in my chat:
```
Access to XMLHttpRequest at 'http://172.17.0.1:9140/ocs/v1.php/cloud/capabilities?format=json' from origin 'http://172.17.0.1:9100' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
core.bundle.js:110 Seems that your token is invalid. Error: Error: CORS request rejected: http://172.17.0.1:9140/ocs/v1.php/cloud/capabilities?format=json
    at l (core.bundle.js:68)
    at XMLHttpRequest.i.onreadystatechange (core.bundle.js:68)
```

Is it related?

AFAICT ocis-reva uses a CORS middleware, which by default allows access from all origins ... see https://github.com/owncloud/ocis-reva/blob/master/pkg/command/frontend.go#L102

**Comment by dpakach** [comment]
> @dpakach how is this blocked? the CI tests are green.
> 

@butonic I have not yet configured drone to run tests with ocis. These are just the old tests running with oc10.


> Is it related?

Yes, that is the main reason we are blocked right now. Since in CI we have to run our browser and selenium inside the docker container, we will have to access phoenix from different host and at the moment, I couldn't do that.

> AFAICT ocis-reva uses a CORS middleware, which by default allows access from all origins ... see https://github.com/owncloud/ocis-reva/blob/master/pkg/command/frontend.go#L102

That seems to be true, In that case the problem might be with with konnectd itself (I'm not sure though). I found some things and wrote about it  [here](https://github.com/owncloud/ocis-konnectd/issues/20#issuecomment-588761089).  Please have a look at it. It might be helpful.

File: .drone.star [hunk 9]
```diff
@@ -74,6 +74,23 @@ config = {
...
+				'all': 'webUIOCIS'
+			},
+			'extraEnvironment': {
+				'SERVER_HOST': 'http://ocis:9100',
+				'BACKEND_HOST': 'http://reva:9140',
+				'RUN_ON_OCIS': 'true',
+				'OCIS_SKELETON_DIR': '/var/www/owncloud/server/apps/testing/data/webUISkeleton',
+				'OCIS_REVA_DATA_ROOT': '/srv/app/tmp/reva/',
+				'LDAP_SERVER_URL': 'ldap://ldap',
+				'CI': 'true',
```
Comment (individual-it) [inline .drone.star]: no need when running in CI, that variable will be always true

File: .drone.star [hunk 10]
```diff
@@ -305,6 +322,7 @@ def acceptance():
 		'logLevel': '2',
 		'federatedServerNeeded': False,
 		'federatedServerVersion': '',
+		'runningOnOCIS': False,
```
Comment (individual-it) [inline .drone.star]: can we set that variable to default false?

File: .drone.star [hunk 11]
```diff
@@ -713,7 +755,7 @@ def installNPM():
 		'image': 'owncloudci/nodejs:10',
 		'pull': 'always',
 		'commands': [
-			'yarn install --frozen-lockfile'
+			'yarn install --frozen-lockfile',
```
Comment (individual-it) [inline .drone.star]: why this change?

File: .drone.star [hunk 12]
```diff
@@ -734,8 +776,15 @@ def buildPhoenix():
 		'pull': 'always',
 		'commands': [
 			'yarn dist',
-			'cp tests/drone/config.json dist/config.json'
-		]
+			'cp tests/drone/config.json dist/config.json',
+			'mkdir -p /srv/config',
+			'cp -r /var/www/owncloud/phoenix/tests/drone /srv/config',
+			'ls -la /srv/config/drone'
```
Comment (individual-it) [inline .drone.star]: extra debug line?

File: .drone.star [hunk 13]
```diff
@@ -980,16 +1191,53 @@ def runWebuiAcceptanceTests(suite, alternateSuiteName, filterTags, extraEnvironm
...
+
 	return [{
 		'name': 'webui-acceptance-tests',
 		'image': 'owncloudci/nodejs:11',
 		'pull': 'always',
 		'environment': environment,
 		'commands': [
 			'cd /var/www/owncloud/phoenix',
-			'curl http://phoenix/oidc-callback.html',
+			# 'curl http://phoenix/oidc-callback.html' if not runningOnOCIS else 'curl http://ocis:9100/oidc-callback.html',
```
Comment (individual-it) [inline .drone.star]: debug line, please remove

File: .drone.star (hunk 10 above)
Comment (dpakach) [reply .drone.star]: Yes, That is exactly what we are doing here.

File: .drone.star (hunk 12 above)
Comment (dpakach) [reply .drone.star]: Not exactly, It would be nice to see if the files are copied in the logs.

File: .drone.star [hunk 14]
```diff
@@ -354,28 +371,46 @@ def acceptance():
...
+									phoenixService() +
+									owncloudService() +
+									(
+										owncloudFederatedService() +
+										databaseServiceForFederation(db, federationDbSuffix)
+										if params['federatedServerNeeded'] else []
+									)
+								) if not params['runningOnOCIS'] else (
+									browserService(alternateSuiteName, browser) +
+									databaseService(db) +
```
Comment (phil-davis) [inline .drone.star]: `browserService` and `databaseService` are duplicated in the `if` and `else`. Those could be put outside this if-else.
